### PR TITLE
feat: add support for symbolic built-in functions and handle built-in…

### PR DESCRIPTION
**Pull request by ajberdy**
… constants symbolically

*Issue #, if available:*

*Description of changes:*

Add support for using built-in functions on symbolic variables and treat built-in constants symbolically

*Testing done:*

unit testing

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/default_simulator/openqasm/_helpers/functions.py [hunk 1]
```diff
@@ -142,55 +144,156 @@
...
 
-builtin_constants = {
-    "pi": FloatLiteral(np.pi),
-    "π": FloatLiteral(np.pi),
-    "tau": FloatLiteral(2 * np.pi),
-    "τ": FloatLiteral(2 * np.pi),
-    "euler": FloatLiteral(np.e),
-    "ℇ": FloatLiteral(np.e),
-}
+class BuiltinFunctions:
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/_helpers/functions.py:157]: fair enough, I could easily move these to a new module. The reason I moved away from the dict is cause the implementation got too complex to justify one-line lambdas, which meant I couldn't define them inline anymore

File: src/braket/default_simulator/openqasm/_helpers/functions.py [hunk 2]
```diff
@@ -142,55 +144,156 @@
...
-    "pi": np.pi,
-    "tau": 2 * np.pi,
-    "euler": np.e,
+builtin_constants = {
+    "pi": SymbolLiteral(sympy.pi),
+    "π": SymbolLiteral(sympy.pi),
+    "tau": SymbolLiteral(2 * sympy.pi),
+    "τ": SymbolLiteral(2 * sympy.pi),
+    "euler": SymbolLiteral(sympy.E),
+    "ℇ": SymbolLiteral(sympy.E),
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/_helpers/functions.py:153]: I think it's to not cause collisions with `a`, `b`, `c`, `d`, `e`, ...

File: src/braket/default_simulator/openqasm/interpreter.py [hunk 3]
```diff
@@ -577,8 +577,8 @@ def _(self, node: FunctionCall) -> Optional[QASMNode]:
         self._uses_advanced_language_features = True
         function_name = node.name.name
         arguments = self.visit(node.arguments)
-        if function_name in builtin_functions:
-            return builtin_functions[function_name](*arguments)
+        if function_name in dir(BuiltinFunctions) and not function_name.startswith("__"):
+            return getattr(BuiltinFunctions, function_name)(*arguments)
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/interpreter.py:581]: I could keep the dict, but I'm not convinced it adds any specific value here. If you feel like it's a better setup, then I'm happy to bring it back though

[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/interpreter.py:581]: what if I moved the BuiltinFunctions to a separate module and replaced the dict with a function `get_builtin_method(name)`. That way the interface wouldn't be meta-programming, rather pretty analogous to a dict look-up, but we could still simplify the implementation with getattr under the hood to avoid manually building a redundant map

[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/default_simulator/openqasm/interpreter.py:581]: Yeah, definitely like that more!

File: src/braket/default_simulator/openqasm/_helpers/casting.py [hunk 4]
```diff
@@ -116,19 +117,21 @@ def _(into: UintType, variable: LiteralType) -> IntegerLiteral:
...
     if into.size is None:
-        return FloatLiteral(float(variable.value))
-    if into.size.value not in (16, 32, 64):
-        raise ValueError("Float size must be one of {16, 32, 64}.")
-    value = float(np.array(variable.value, dtype=np.dtype(f"float{into.size.value}")))
+        value = float(value)
+    else:
+        if into.size.value not in (16, 32, 64):
+            raise ValueError("Float size must be one of {16, 32, 64}.")
+        value = float(np.array(value, dtype=np.dtype(f"float{into.size.value}")))
```
Comment (speller26) [inline src/braket/default_simulator/openqasm/_helpers/casting.py:126]: Why the change back here?

**Review by speller26 — APPROVED** [review]

File: src/braket/default_simulator/openqasm/_helpers/casting.py (hunk 4 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/_helpers/casting.py:126]: hadn't pulled in the changes when I started this code changes and `variable.value` -> `value` created a conflict and I ended up just selecting one half of the diff to resolve. I can change back, but I also don't mind factoring out the common logic of wrapping in a FloatLiteral as written
